ext/spl/spl_fixedarray.c

```diff
@@ -106,15 +106,17 @@ static void spl_fixedarray_resize(spl_fixedarray *array, zend_long size) /* {{{
 
 	/* clearing the array */
 	if (size == 0) {
-		zend_long i;
-
-		for (i = 0; i < array->size; i++) {
-			zval_ptr_dtor(&(array->elements[i]));
-		}
+		if (array->elements != NULL) {
+			zend_long i;
+			zval *elements = array->elements;
 
-		if (array->elements) {
-			efree(array->elements);
 			array->elements = NULL;
+
+			for (i = 0; i < array->size; i++) {
+				zval_ptr_dtor(&(elements[i]));
+			}
+
+			efree(elements);
 		}
```

Comment:
Actually, there seem like there may be good reasons to keep this optimization for size 0 in narrow edge cases - e.g. if a PHP program was using virtually all of its memory limit for a single SplFixedArray, this would exceed it for certain ways of resetting the array.

But still, I realized that there's another bug - if you call offsetGet() or offsetSet() from an element destructor, it will be in range because the size hasn't been set to 0 yet, but this will be writing to an offset relative to null, either segfaulting or (still having the problem of) corrupting memory if the array is large enough

The below amendment would make setting the size to 0 work

```suggestion
			array->elements = NULL;
			array->size = 0;

			for (i = 0; i < array->size; i++) {
				zval_ptr_dtor(&(elements[i]));
			}

			efree(elements);
			return;
		}
```

Reply:
Oops, I misread that as meaning `==` in an if statement.